# MC316 Final Exam – Part B: PYTHON

* You will have 1 hour and 40 minutes to finish the test.
* This part includes 10 questions. Each question is worth 15 points.
* You get extra 30 points for full answers.
* Please read the instructions carefully. 

## I. SET UP


### Import pandas library for use in scripting. 

In [1]:
import pandas as pd

### Import the datasets and name them.
For this test, you will analyze revenue and expenditure data released by Kansas Department of Administration for the fiscal year 2018. While the original data can be directly downloaded [here](http://kanview.ks.gov/DataDownload.aspx), for your convenience, the instructor had them uploaded to a server at the following urls. 

#### For Kansas State University:
1. Expenditure data: https://datajournalism.tech/wp-content/uploads/2019/05/ExpData_367_2018.csv
2. Revenue data: https://datajournalism.tech/wp-content/uploads/2019/05/RevData_367_2018.csv

#### For University of Kansas:
3. Expenditure data: https://datajournalism.tech/wp-content/uploads/2019/05/ExpData_682_2018.csv
4. Revenue data: https://datajournalism.tech/wp-content/uploads/2019/05/RevData_682_2018.csv

You will need to import all 4 datasets and name them in an easy way to call out later.

*Please refer to the [Glossary](http://kanview.ks.gov/Glossary.aspx#f) for data description.

In [2]:
KSUexpenditure = pd.read_csv("https://datajournalism.tech/wp-content/uploads/2019/05/ExpData_367_2018.csv")

In [44]:
KSUexpenditure.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 24094 entries, 0 to 24093
Data columns (total 22 columns):
Fiscal_Year                        24094 non-null int64
Fiscal_Quarter                     24094 non-null int64
Business_Unit                      24094 non-null int64
Business_Unit_Descr                24094 non-null object
GAAP_Fund_Type                     24094 non-null int64
GAAP_Fund_Type_Descr               24094 non-null object
Fund_Code                          24094 non-null int64
Fund_Code_Descr                    24094 non-null object
Budget_Ref                         24094 non-null int64
Budget_Ref_Descr                   24094 non-null object
Program_Code                       24094 non-null int64
Program_Code_Descr                 24094 non-null object
State_Function                     24094 non-null int64
State_Function_Descr               24094 non-null object
Primary_Account_Code               24094 non-null int64
Primary_Account_Code_Descr         24094 non-

In [3]:
KSUrevenue = pd.read_csv("https://datajournalism.tech/wp-content/uploads/2019/05/RevData_367_2018.csv")

In [45]:
KSUrevenue.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1910 entries, 0 to 1909
Data columns (total 22 columns):
Fiscal_Year                        1910 non-null int64
Fiscal_Quarter                     1910 non-null int64
Business_Unit                      1910 non-null int64
Business_Unit_Descr                1910 non-null object
GAAP_Fund_Type                     1910 non-null int64
GAAP_Fund_Type_Descr               1910 non-null object
Fund_Code                          1910 non-null int64
Fund_Code_Descr                    1910 non-null object
Budget_Ref                         1910 non-null int64
Budget_Ref_Descr                   1910 non-null object
Program_Code                       1910 non-null int64
Program_Code_Descr                 1910 non-null object
State_Function                     1910 non-null int64
State_Function_Descr               1910 non-null object
Primary_Account_Code               1910 non-null int64
Primary_Account_Code_Descr         1910 non-null object
Interm

In [4]:
KUexpenditure = pd.read_csv("https://datajournalism.tech/wp-content/uploads/2019/05/ExpData_682_2018.csv")

In [5]:
KUrevenue = pd.read_csv("https://datajournalism.tech/wp-content/uploads/2019/05/RevData_682_2018.csv")

## II. REQUIRED QUESTIONS
* Make sure you type the answers in the cells right below the codes.
* If you need to add extra cells, click the + button in the toolbar up the top. 

### Q1: What is the total revenue recorded for Kansas State University in the fiscal year of 2018? How does it compare to the total revenue recorded for University of Kansas in the same year?

In [20]:
KSUrevenue.Total.sum()

854324643.84

In [21]:
KUrevenue.Total.sum()

808767902.55

In [22]:
854324643.84-808767902.55

45556741.29000008

<b>Answer:</b> KSU's total 2018 revenue was 854,324,643.84 and KU's was 808,767,902.55 and the KSU had 45556741.29 more than KU.

### Q2: What is the total expense recorded for Kansas State University in the fiscal year of 2018? How does it compare to the total expense recorded for University of Kansas in the same year?

In [25]:
KSUexpenditure.Total.sum()

1018584373.02

In [26]:
KUexpenditure.Total.sum()

990238875.6099999

In [27]:
1018584373.02-990238875.6099999

28345497.410000086

<b>Answer:</b> KSU's total expense record was 1,018,584,373.02 and KU's total expense record was 990,238,875.61 and the KSU's was greater by 28,345,497.41.

### Q3: Where did the two universities' primary revenues come from? 
*Refer to the description [here](http://kanview.ks.gov/Revenues/AcctCodeRevenues.aspx) to write your answer.

In [41]:
KSUrevenue.sort_values("Total", ascending=False).head()

Fiscal_Year  Fiscal_Quarter  Business_Unit  \
2            2018               1            367   
944          2018               3            367   
1255         2018               3            367   
301          2018               1            367   
489          2018               2            367   

                 Business_Unit_Descr  GAAP_Fund_Type  \
2     Kansas State University                     12   
944   Kansas State University                     12   
1255  Kansas State University                     12   
301   Kansas State University                     12   
489   Kansas State University                     12   

                GAAP_Fund_Type_Descr  Fund_Code  \
2     Component Unit                       2062   
944   Component Unit                       2062   
1255  Component Unit                       3142   
301   Component Unit                       3142   
489   Component Unit                       2062   

                     Fund_Code_Descr  Budget_Ref  \
2     GENERAL FF                            2000   
944   GENERAL FF                            2000   
1255  UNIVERSITY FDF                        3146   
301   UNIVERSITY FDF                        3146   
489   GENERAL FF                            2000   

                                       Budget_Ref_Descr  ...  State_Function  \
2     GENERAL FF                                    ...  ...              26   
944   GENERAL FF                                    ...  ...              26   
1255  UNI FDF--STUDENT FINANCIAL AID                ...  ...              26   
301   UNI FDF--STUDENT FINANCIAL AID                ...  ...              26   
489   GENERAL FF                                    ...  ...              26   

                                   State_Function_Descr  Primary_Account_Code  \
2     Universities                                  ...                  4200   
944   Universities                                  ...                  4200   
1255  Universities                                  ...                  4400   
301   Universities                                  ...                  4400   
489   Universities                                  ...                  4200   

          Primary_Account_Code_Descr  Intermediate_Account_Code  \
2     AGENCY EARNINGS                                     42000   
944   AGENCY EARNINGS                                     42000   
1255  GIFT DONATION FEDERAL GRANTS                        44000   
301   GIFT DONATION FEDERAL GRANTS                        44000   
489   AGENCY EARNINGS                                     42000   

     Intermediate_Account_Code_Descr  Detail_Account_Code  \
2     SERVICE CHARGES                              420500   
944   SERVICE CHARGES                              420500   
1255  FEDERAL GRANTS                               440100   
301   FEDERAL GRANTS                               440100   
489   SERVICE CHARGES                              420500   

           Detail_Account_Code_Descr        Total Unnamed: 21  
2     EDUCATION AND LIBRARIES        91090183.660         nan  
944   EDUCATION AND LIBRARIES        90454678.170         nan  
1255  FEDERAL GRANT OPERATING        65000933.550         nan  
301   FEDERAL GRANT OPERATING        62940038.200         nan  
489   EDUCATION AND LIBRARIES        15659557.110         nan  

[5 rows x 22 columns]

In [117]:
KUrevenue.Primary_Account_Code_Descr.sort_values()

0       AGENCY EARNINGS               
585     AGENCY EARNINGS               
584     AGENCY EARNINGS               
583     AGENCY EARNINGS               
582     AGENCY EARNINGS               
581     AGENCY EARNINGS               
580     AGENCY EARNINGS               
574     AGENCY EARNINGS               
573     AGENCY EARNINGS               
572     AGENCY EARNINGS               
567     AGENCY EARNINGS               
566     AGENCY EARNINGS               
565     AGENCY EARNINGS               
564     AGENCY EARNINGS               
563     AGENCY EARNINGS               
562     AGENCY EARNINGS               
561     AGENCY EARNINGS               
554     AGENCY EARNINGS               
517     AGENCY EARNINGS               
520     AGENCY EARNINGS               
525     AGENCY EARNINGS               
531     AGENCY EARNINGS               
532     AGENCY EARNINGS               
537     AGENCY EARNINGS               
586     AGENCY EARNINGS               
1075    AGENCY EARNINGS  

<b>Answer:</b> Annual Revenues represents agency receipts and deposits into funds established within the state treasury, including: taxes, agency earnings, revenue from the use of money and property, gifts, donations, federal grants, and other revenues.


### Q4: What was the biggest source of revenues for Kansas State University? Compare to the one of University of Kansas.

Remove hashtag and run the following code to suppress scientific notation:

In [36]:
pd.set_option('display.float_format', lambda x: '%.3f' % x) 

In [55]:
KSUrevenue.groupby(["Primary_Account_Code_Descr", "Primary_Account_Code"]).sum().sort_values("Total", ascending=False)

Fiscal_Year  \
Primary_Account_Code_Descr     Primary_Account_Code                        
AGENCY EARNINGS                AGENCY EARNINGS                   1285466   
GIFT DONATION FEDERAL GRANTS   GIFT DONATION FEDERAL GRANTS       306736   
NON REVENUE RECEIPTS           NON REVENUE RECEIPTS               522662   
REVENUE TRANSFERS              REVENUE TRANSFERS                  688138   
REV FROM USE OF MONEY & PROP   REV FROM USE OF MONEY & PROP       714372   
OTHER REVENUE                  OTHER REVENUE                      322880   
TAXES                          TAXES                               14126   

                                                             Fiscal_Quarter  \
Primary_Account_Code_Descr     Primary_Account_Code                           
AGENCY EARNINGS                AGENCY EARNINGS                         1598   
GIFT DONATION FEDERAL GRANTS   GIFT DONATION FEDERAL GRANTS             377   
NON REVENUE RECEIPTS           NON REVENUE RECEIPTS                     666   
REVENUE TRANSFERS              REVENUE TRANSFERS                        827   
REV FROM USE OF MONEY & PROP   REV FROM USE OF MONEY & PROP             883   
OTHER REVENUE                  OTHER REVENUE                            409   
TAXES                          TAXES                                     21   

                                                             Business_Unit  \
Primary_Account_Code_Descr     Primary_Account_Code                          
AGENCY EARNINGS                AGENCY EARNINGS                      233779   
GIFT DONATION FEDERAL GRANTS   GIFT DONATION FEDERAL GRANTS          55784   
NON REVENUE RECEIPTS           NON REVENUE RECEIPTS                  95053   
REVENUE TRANSFERS              REVENUE TRANSFERS                    125147   
REV FROM USE OF MONEY & PROP   REV FROM USE OF MONEY & PROP         129918   
OTHER REVENUE                  OTHER REVENUE                         58720   
TAXES                          TAXES                                  2569   

                                                             GAAP_Fund_Type  \
Primary_Account_Code_Descr     Primary_Account_Code                           
AGENCY EARNINGS                AGENCY EARNINGS                         7615   
GIFT DONATION FEDERAL GRANTS   GIFT DONATION FEDERAL GRANTS            1824   
NON REVENUE RECEIPTS           NON REVENUE RECEIPTS                    3108   
REVENUE TRANSFERS              REVENUE TRANSFERS                       4092   
REV FROM USE OF MONEY & PROP   REV FROM USE OF MONEY & PROP            4182   
OTHER REVENUE                  OTHER REVENUE                           1920   
TAXES                          TAXES                                     84   

                                                             Fund_Code  \
Primary_Account_Code_Descr     Primary_Account_Code                      
AGENCY EARNINGS                AGENCY EARNINGS                 2115732   
GIFT DONATION FEDERAL GRANTS   GIFT DONATION FEDERAL GRANTS     442889   
NON REVENUE RECEIPTS           NON REVENUE RECEIPTS             913600   
REVENUE TRANSFERS              REVENUE TRANSFERS               1612445   
REV FROM USE OF MONEY & PROP   REV FROM USE OF MONEY & PROP    1966220   
OTHER REVENUE                  OTHER REVENUE                    531619   
TAXES                          TAXES                             17640   

                                                             Budget_Ref  \
Primary_Account_Code_Descr     Primary_Account_Code                       
AGENCY EARNINGS                AGENCY EARNINGS                  1900362   
GIFT DONATION FEDERAL GRANTS   GIFT DONATION FEDERAL GRANTS      362768   
NON REVENUE RECEIPTS           NON REVENUE RECEIPTS              820788   
REVENUE TRANSFERS              REVENUE TRANSFERS                1497807   
REV FROM USE OF MONEY & PROP   REV FROM USE OF MONEY & PROP     1806495   
OTHER REVENUE                  OTHER REVENUE             

In [57]:
KUrevenue.groupby(["Primary_Account_Code_Descr", "Primary_Account_Code"]).sum().sort_values("Total", ascending=False)

Fiscal_Year  \
Primary_Account_Code_Descr     Primary_Account_Code                
AGENCY EARNINGS                4200                       758768   
GIFT DONATION FEDERAL GRANTS   4400                        76684   
NON REVENUE RECEIPTS           4600                       316826   
OTHER REVENUE                  4500                       115026   
REV FROM USE OF MONEY & PROP   4300                       506518   
REVENUE TRANSFERS              7600                       308754   
TAXES                          4100                        90810   

                                                     Fiscal_Quarter  \
Primary_Account_Code_Descr     Primary_Account_Code                   
AGENCY EARNINGS                4200                            1010   
GIFT DONATION FEDERAL GRANTS   4400                              94   
NON REVENUE RECEIPTS           4600                             399   
OTHER REVENUE                  4500                             140   
REV FROM USE OF MONEY & PROP   4300                             621   
REVENUE TRANSFERS              7600                             404   
TAXES                          4100                             114   

                                                     Business_Unit  \
Primary_Account_Code_Descr     Primary_Account_Code                  
AGENCY EARNINGS                4200                         256432   
GIFT DONATION FEDERAL GRANTS   4400                          25916   
NON REVENUE RECEIPTS           4600                         107074   
OTHER REVENUE                  4500                          38874   
REV FROM USE OF MONEY & PROP   4300                         171182   
REVENUE TRANSFERS              7600                         104346   
TAXES                          4100                          30690   

                                                     GAAP_Fund_Type  \
Primary_Account_Code_Descr     Primary_Account_Code                   
AGENCY EARNINGS                4200                            4512   
GIFT DONATION FEDERAL GRANTS   4400                             456   
NON REVENUE RECEIPTS           4600                            1884   
OTHER REVENUE                  4500                             684   
REV FROM USE OF MONEY & PROP   4300                            3012   
REVENUE TRANSFERS              7600                            1836   
TAXES                          4100                             540   

                                                     Fund_Code  Budget_Ref  \
Primary_Account_Code_Descr     Primary_Account_Code                          
AGENCY EARNINGS                4200                    1343448     1234420   
GIFT DONATION FEDERAL GRANTS   4400                     117009      109001   
NON REVENUE RECEIPTS           4600                     545872      493440   
OTHER REVENUE                  4500                     170273      152085   
REV FROM USE OF MONEY & PROP   4300                    1407791     1354925   
REVENUE TRANSFERS              7600                     787108      755368   
TAXES                          4100                     141643      130460   

                                                     Program_Code  \
Primary_Account_Code_Descr     Primary_Account_Code                 
AGENCY EARNINGS                4200                      18742000   
GIFT DONATION FEDERAL GRANTS   4400                       1736000   
NON REVENUE RECEIPTS           4600                       8553000   
OTHER REVENUE                  4500                       2510000   
REV FROM USE OF MONEY & PROP   4300                      15262000   
REVENUE TRANSFERS              7600                      10139000   
TAXES                          4100                       2260000   

                                                     State_Function  \
Primary_Account_Code_Descr     Primary_Account_Code                   
AGENCY EARNINGS                4200                           

In [58]:
412565112.820 - 408409337.320

4155775.5

<b>Answer:</b> KSU's biggest source of revenue is 412565112.820	 from agency earnings and KU's biggest source is 408409337.320 from agency earnings. KSU's is bigger by 4155775.5



### Q5: What accounts contributed to Kansas State University's biggest source of revenues?
If you run into a situation where a filter function returns 0 results, you may want to check if there are leading and trailing spaces in the column. To clean the leading and trailing spaces, you will need to remove the hashtag, revise the names, and run the following code:

In [76]:
KSUrevenue["Detail_Account_Code"] = KSUrevenue["Detail_Account_Code_Descr"].str.strip()

After cleaning the spaces, you can start filtering your datasets to find the answer for Q5.

In [77]:
KSUrevenue.Detail_Account_Code.value_counts()

AVERAGE DAILY BALANCE INTEREST    232
TECHNICAL AND SKILLED SERVICES    117
OTHER MISCELLANEOUS REVENUE       114
EDUCATION AND LIBRARIES           112
OTHER SERVICE CHARGES             110
DEPARTMENTAL OR AGENCY SALES       99
OTHER NONREVENUE RECEIPTS          98
OPERATING TRANSFERS IN             85
ALL OTHER OPERATING GRANTS         81
OPERATING TRANSFERS OUT            76
OP TRANSFER OUT INTEREST ALLOC     75
FEDERAL GRANT OPERATING            71
OTHER REIMB AND REFUNDS            61
OTHER RENTS AND ROYALTIES          54
OTHER COMMODITIES                  47
AGY FUNDS AUTHORIZED DEBT SVC      44
OTHER FINES PENALTIES FORFEIT      42
OP TRANSFER IN INTEREST ALLOC      38
RECOVERY OF CURRENT FY EXP         37
RENT HALLS & ROOMS ST BLDGS        35
SUSPENSE                           25
FED SUBGRANT TRANSFER IN           23
MEALS AND PROCESSED FOODS          23
SALVAGED MATERIALS                 18
CLERICAL SERVICES                  16
OTHER INTEREST DIVIDEND PREM       15
OTHER ADMISS

In [118]:
KSUrevenue.groupby(["Primary_Account_Code_Descr", "Detail_Account_Code", "Detail_Account_Code_Descr"]).sum().sort_values("Total", ascending=False)

Fiscal_Year  \
Primary_Account_Code_Descr     Detail_Account_Code            Detail_Account_Code_Descr                     
AGENCY EARNINGS                EDUCATION AND LIBRARIES        EDUCATION AND LIBRARIES              226016   
GIFT DONATION FEDERAL GRANTS   FEDERAL GRANT OPERATING        FEDERAL GRANT OPERATING              143278   
                               ALL OTHER OPERATING GRANTS     ALL OTHER OPERATING GRANTS           163458   
AGENCY EARNINGS                DEPARTMENTAL OR AGENCY SALES   DEPARTMENTAL OR AGENCY SALES         199782   
                               TECHNICAL AND SKILLED SERVICES TECHNICAL AND SKILLED SERVICES       236106   
REVENUE TRANSFERS              OPERATING TRANSFERS IN         OPERATING TRANSFERS IN               171530   
                               AGY FUNDS AUTHORIZED DEBT SVC  AGY FUNDS AUTHORIZED DEBT SVC         88792   
REV FROM USE OF MONEY & PROP   RENT HALLS & ROOMS ST BLDGS    RENT HALLS & ROOMS ST BLDGS           70630   
NON REVENUE RECEIPTS           OTHER REIMB AND REFUNDS        OTHER REIMB AND REFUNDS              123098   
AGENCY EARNINGS                MEALS AND PROCESSED FOODS      MEALS AND PROCESSED FOODS             46414   
NON REVENUE RECEIPTS           OTHER NONREVENUE RECEIPTS      OTHER NONREVENUE RECEIPTS            197764   
                               RECOVERY OF CURRENT FY EXP     RECOVERY OF CURRENT FY EXP            74666   
                               INSURANCE REIMBURSEMENTS       INSURANCE REIMBURSEMENTS               2018   
AGENCY EARNINGS                OTHER SERVICE CHARGES          OTHER SERVICE CHARGES                221980   
NON REVENUE RECEIPTS           REVENUE BOND ISSUE AT PAR      REVENUE BOND ISSUE AT PAR              4036   
REVENUE TRANSFERS              FED SUBGRANT TRANSFER IN       FED SUBGRANT TRANSFER IN              46414   
OTHER REVENUE                  OTHER MISCELLANEOUS REVENUE    OTHER MISCELLANEOUS REVENUE          230052   
REV FROM USE OF MONEY & PROP   OTHER RENTS AND ROYALTIES      OTHER RENTS AND ROYALTIES            108972   
AGENCY EARNINGS                FARM AND DAIRY PRODUCTS        FARM AND DAIRY PRODUCTS               16144   
NON REVENUE RECEIPTS           REIMB FROM OTHER ST AGENCY     REIMB FROM OTHER ST AGENCY            24216   
AGENCY EARNINGS                LIVESTOCK AND POULTRY          LIVESTOCK AND POULTRY                 14126   
NON REVENUE RECEIPTS           SALE OF FIXED ASSETS           SALE OF FIXED ASSETS                   4036   
AGENCY EARNINGS                CARE AND HOSPITALIZATION OTHER CARE AND HOSPITALIZATION OTHER         8072   
OTHER REVENUE                  OTHER FINES PENALTIES FORFEIT  OTHER FINES PENALTIES FORFEIT         84756   
AGENCY EARNINGS                ADMISSIONS TO PLAYS & CONCERTS ADMISSIONS TO PLAYS & CONCERTS        14126   
                               OTHER COMMODITIES              OTHER COMMODITIES                     94846   
REV FROM USE OF MONEY & PROP   AVERAGE DAILY BALANCE INTEREST AVERAGE DAILY BALANCE INTEREST       468176   
REVENUE TRANSFERS              OP TRANSFER IN INTEREST ALLOC  OP TRANSFER IN INTEREST ALLOC         76684   
AGENCY EARNINGS                CORPORATION SERVICES           CORPORATION SERVICES                  22198   
                               CLERICAL SERVICES              CLERICAL SERVICES                     32288   
REV FROM USE OF MONEY & PROP   OTHER INTEREST DIVIDEND PREM   OTHER INTEREST DIVIDEND PREM          30270   
AGENCY EARNINGS                USABLE CONDEMNED EQUIPMENT     USABLE CONDEMNED EQUIPMENT            18162   
                               STATE PUBLICATIONS             STATE PUBLICATIONS                    26234   
                               AUDITS AND INSPECTIONS         AUDITS AND INSPECTIONS                10090   
                               OTHER ADMISSIONS               OTHER ADMISSIONS                      30270   
                               SALVAGED MATERIALS            

<b>Answer:</b> Agency Earnings contributed the most to KSU's biggest source of revenues then Gift Donation Federal Grants and specifically Education and Libraries, meals and processed foods, other service charges, farm and dairy products, livestock and poultry, Admissions to plays and concerts, corporation services, usable condemned equipment, manufactured products.

### Q6: Where did the money go? (Answer for both universities)
*Refer to the description [here] (http://kanview.ks.gov/Expenditures/AcctCodeExpenditures.aspx) to write your answer.

In [51]:
KSUexpenditure

Fiscal_Year  Fiscal_Quarter  Business_Unit  \
0             2018               1            367   
1             2018               1            367   
2             2018               1            367   
3             2018               1            367   
4             2018               1            367   
5             2018               1            367   
6             2018               1            367   
7             2018               1            367   
8             2018               1            367   
9             2018               1            367   
10            2018               1            367   
11            2018               1            367   
12            2018               1            367   
13            2018               1            367   
14            2018               1            367   
15            2018               1            367   
16            2018               1            367   
17            2018               1            367   
18            2018               1            367   
19            2018               1            367   
20            2018               1            367   
21            2018               1            367   
22            2018               1            367   
23            2018               1            367   
24            2018               1            367   
25            2018               1            367   
26            2018               1            367   
27            2018               1            367   
28            2018               1            367   
29            2018               1            367   
...            ...             ...            ...   
24064         2018               4            367   
24065         2018               4            367   
24066         2018               4            367   
24067         2018               4            367   
24068         2018               4            367   
24069         2018               4            367   
24070         2018               4            367   
24071         2018               4            367   
24072         2018               4            367   
24073         2018               4            367   
24074         2018               4            367   
24075         2018               4            367   
24076         2018               4            367   
24077         2018               4            367   
24078         2018               4            367   
24079         2018               4            367   
24080         2018               4            367   
24081         2018               4            367   
24082         2018               4            367   
24083         2018               4            367   
24084         2018               4            367   
24085         2018               4            367   
24086         2018               4            367   
24087         2018               4            367   
24088         2018               4            367   
24089         2018               4            367   
24090         2018               4            367   
24091         2018               4            367   
24092         2018               4            367   
24093         2018               4            367   

                  Business_Unit_Descr  GAAP_Fund_Type  \
0      Kansas State University                     12   
1      Kansas State University                     12   
2      Kansas State University                     12   
3      Kansas State University                     12   
4      Kansas State University                     12   
5      Kansas State University                     12   
6      Kansas State University                     12   
7      Kansas State University                     12   
8      Kansas State University                     12   
9      Kansas State University                     12   
10     Kansas State University                     12   
11     Kansas State University      

In [52]:
KUexpenditure

Fiscal_Year  Fiscal_Quarter  Business_Unit  \
0             2018               1            682   
1             2018               1            682   
2             2018               1            682   
3             2018               1            682   
4             2018               1            682   
5             2018               1            682   
6             2018               1            682   
7             2018               1            682   
8             2018               1            682   
9             2018               1            682   
10            2018               1            682   
11            2018               1            682   
12            2018               1            682   
13            2018               1            682   
14            2018               1            682   
15            2018               1            682   
16            2018               1            682   
17            2018               1            682   
18            2018               1            682   
19            2018               1            682   
20            2018               1            682   
21            2018               1            682   
22            2018               1            682   
23            2018               1            682   
24            2018               1            682   
25            2018               1            682   
26            2018               1            682   
27            2018               1            682   
28            2018               1            682   
29            2018               1            682   
...            ...             ...            ...   
10523         2018               4            682   
10524         2018               4            682   
10525         2018               4            682   
10526         2018               4            682   
10527         2018               4            682   
10528         2018               4            682   
10529         2018               4            682   
10530         2018               4            682   
10531         2018               4            682   
10532         2018               4            682   
10533         2018               4            682   
10534         2018               4            682   
10535         2018               4            682   
10536         2018               4            682   
10537         2018               4            682   
10538         2018               4            682   
10539         2018               4            682   
10540         2018               4            682   
10541         2018               4            682   
10542         2018               4            682   
10543         2018               4            682   
10544         2018               4            682   
10545         2018               4            682   
10546         2018               4            682   
10547         2018               4            682   
10548         2018               4            682   
10549         2018               4            682   
10550         2018               4            682   
10551         2018               4            682   
10552         2018               4            682   

                  Business_Unit_Descr  GAAP_Fund_Type  \
0      University of Kansas                         1   
1      University of Kansas                         1   
2      University of Kansas                         1   
3      University of Kansas                         1   
4      University of Kansas                         1   
5      University of Kansas                         1   
6      University of Kansas                         1   
7      University of Kansas                         1   
8      University of Kansas                         1   
9      University of Kansas                         1   
10     University of Kansas                         1   
11     University of Kansas         

<b>Answer:</b> Salaries and Wages, Contractural Services, Commodities, Capital Outlay, Grants, Claims and Shared Revenue, Debt Services, and Non-Expense Items.

### Q7: What costs two universities the most? 

In [59]:
KSUexpenditure.groupby(["Primary_Account_Code_Descr", "Primary_Account_Code"]).sum().sort_values("Total", ascending=False)

Fiscal_Year  \
Primary_Account_Code_Descr     Primary_Account_Code                
SALARIES AND WAGES             5100                     15700040   
NONEXPENSE ITEMS               5700                       260322   
CONTRACTUAL SERVICES           5200                     21312098   
GRANTS CLAIMS SHARED REVENUE   5500                      1077612   
CAPITAL OUTLAY                 5400                      2367114   
COMMODITIES                    5300                      7547320   
EXPENSE TRANSFERS              7700                       213908   
DEBT SERVICE                   5600                       143278   

                                                     Fiscal_Quarter  \
Primary_Account_Code_Descr     Primary_Account_Code                   
SALARIES AND WAGES             5100                           19631   
NONEXPENSE ITEMS               5700                             313   
CONTRACTUAL SERVICES           5200                           26780   
GRANTS CLAIMS SHARED REVENUE   5500                            1344   
CAPITAL OUTLAY                 5400                            2990   
COMMODITIES                    5300                            9510   
EXPENSE TRANSFERS              7700                             274   
DEBT SERVICE                   5600                             178   

                                                     Business_Unit  \
Primary_Account_Code_Descr     Primary_Account_Code                  
SALARIES AND WAGES             5100                        2855260   
NONEXPENSE ITEMS               5700                          47343   
CONTRACTUAL SERVICES           5200                        3875887   
GRANTS CLAIMS SHARED REVENUE   5500                         195978   
CAPITAL OUTLAY                 5400                         430491   
COMMODITIES                    5300                        1372580   
EXPENSE TRANSFERS              7700                          38902   
DEBT SERVICE                   5600                          26057   

                                                     GAAP_Fund_Type  \
Primary_Account_Code_Descr     Primary_Account_Code                   
SALARIES AND WAGES             5100                           69809   
NONEXPENSE ITEMS               5700                            1548   
CONTRACTUAL SERVICES           5200                          118141   
GRANTS CLAIMS SHARED REVENUE   5500                            6342   
CAPITAL OUTLAY                 5400                           13186   
COMMODITIES                    5300                           41934   
EXPENSE TRANSFERS              7700                            1185   
DEBT SERVICE                   5600                             852   

                                                     Fund_Code  Budget_Ref  \
Primary_Account_Code_Descr     Primary_Account_Code                          
SALARIES AND WAGES             5100                   18651058    16460398   
NONEXPENSE ITEMS               5700                     450027      414569   
CONTRACTUAL SERVICES           5200                   29101777    26369491   
GRANTS CLAIMS SHARED REVENUE   5500                    1487948     1289771   
CAPITAL OUTLAY                 5400                    3613098     3346996   
COMMODITIES                    5300                   10777809     9619498   
EXPENSE TRANSFERS              7700                     524600      462100   
DEBT SERVICE                   5600                     508417      487621   

                                                     Program_Code  \
Primary_Account_Code_Descr     Primary_Account_Code                 
SALARIES AND WAGES             5100                     368901849   
NONEXPENSE ITEMS               5700                       6144904   
CONTRACTUAL SERVICES           5200                     504375573   
GRANTS CLAIMS SHARED REVENUE   5500                      23950069   
CAPITAL OUTLAY                 5400                   

In [60]:
KUexpenditure.groupby(["Primary_Account_Code_Descr", "Primary_Account_Code"]).sum().sort_values("Total", ascending=False)

Fiscal_Year  \
Primary_Account_Code_Descr     Primary_Account_Code                
SALARIES AND WAGES             5100                      7311214   
NONEXPENSE ITEMS               5700                       163458   
CONTRACTUAL SERVICES           5200                      8721796   
GRANTS CLAIMS SHARED REVENUE   5500                       151350   
CAPITAL OUTLAY                 5400                      1487266   
EXPENSE TRANSFERS              7700                       133188   
DEBT SERVICE                   5600                        88792   
COMMODITIES                    5300                      3238890   

                                                     Fiscal_Quarter  \
Primary_Account_Code_Descr     Primary_Account_Code                   
SALARIES AND WAGES             5100                            9283   
NONEXPENSE ITEMS               5700                             203   
CONTRACTUAL SERVICES           5200                           10945   
GRANTS CLAIMS SHARED REVENUE   5500                             187   
CAPITAL OUTLAY                 5400                            2035   
EXPENSE TRANSFERS              7700                             184   
DEBT SERVICE                   5600                             136   
COMMODITIES                    5300                            4102   

                                                     Business_Unit  \
Primary_Account_Code_Descr     Primary_Account_Code                  
SALARIES AND WAGES             5100                        2470886   
NONEXPENSE ITEMS               5700                          55242   
CONTRACTUAL SERVICES           5200                        2947604   
GRANTS CLAIMS SHARED REVENUE   5500                          51150   
CAPITAL OUTLAY                 5400                         502634   
EXPENSE TRANSFERS              7700                          45012   
DEBT SERVICE                   5600                          30008   
COMMODITIES                    5300                        1094610   

                                                     GAAP_Fund_Type  \
Primary_Account_Code_Descr     Primary_Account_Code                   
SALARIES AND WAGES             5100                           35727   
NONEXPENSE ITEMS               5700                             972   
CONTRACTUAL SERVICES           5200                           49580   
GRANTS CLAIMS SHARED REVENUE   5500                             889   
CAPITAL OUTLAY                 5400                            8418   
EXPENSE TRANSFERS              7700                             697   
DEBT SERVICE                   5600                             528   
COMMODITIES                    5300                           18545   

                                                     Fund_Code  Budget_Ref  \
Primary_Account_Code_Descr     Primary_Account_Code                          
SALARIES AND WAGES             5100                    9712276     8513430   
NONEXPENSE ITEMS               5700                     342779      317050   
CONTRACTUAL SERVICES           5200                   11926990    10836327   
GRANTS CLAIMS SHARED REVENUE   5500                     210791      190993   
CAPITAL OUTLAY                 5400                    2176648     2007885   
EXPENSE TRANSFERS              7700                     260467      245396   
DEBT SERVICE                   5600                     314033      312276   
COMMODITIES                    5300                    4468157     4088596   

                                                     Program_Code  \
Primary_Account_Code_Descr     Primary_Account_Code                 
SALARIES AND WAGES             5100                     180808000   
NONEXPENSE ITEMS               5700                       4041000   
CONTRACTUAL SERVICES           5200                     219810000   
GRANTS CLAIMS SHARED REVENUE   5500                       3428000   
CAPITAL OUTLAY                 5400                   

<b>Answer:</b> Salaries and Wages

### Q8: Calculate the percentage of this cost over each university's total expense.

In [61]:
KSUexpenditure.Total.sum()

1018584373.02

In [63]:
476367053.010/1018584373.02

0.4676755952946929

In [62]:
KUexpenditure.Total.sum()

990238875.6099999

In [64]:
471996016.150/990238875.6099999

0.47664864284311637

<b>Answer:</b> KSU's is 46.77% and KU's is 47.66%

### Q9: Breakdown the cost you found in Q7 for both universities.
Again, if you run into a situation where a filter function returns 0 results, you may want to check if there are leading and trailing spaces in the column. To clean the leading and trailing spaces, you will need to remove the hashtag, revise the names, and run the following code:

In [119]:
KSUexpenditure["Primary_Account_Code_Descr"] = KSUexpenditure["Primary_Account_Code"].str.strip()

In [120]:
KSUexpenditure.Primary_Account_Code_Descr.value_counts()

CONTRACTUAL SERVICES            10561
SALARIES AND WAGES               7780
COMMODITIES                      3740
CAPITAL OUTLAY                   1173
GRANTS CLAIMS SHARED REVENUE      534
NONEXPENSE ITEMS                  129
EXPENSE TRANSFERS                 106
DEBT SERVICE                       71
Name: Primary_Account_Code_Descr, dtype: int64

In [122]:
KSUexpen = KSUexpenditure[KSUexpenditure.Primary_Account_Code_Descr == 'SALARIES AND WAGES']

In [132]:
KSUexpen.head()

Fiscal_Year  Fiscal_Quarter  Business_Unit  \
16         2018               1            367   
17         2018               1            367   
18         2018               1            367   
19         2018               1            367   
20         2018               1            367   

               Business_Unit_Descr  GAAP_Fund_Type  \
16  Kansas State University                     12   
17  Kansas State University                     12   
18  Kansas State University                     12   
19  Kansas State University                     12   
20  Kansas State University                     12   

              GAAP_Fund_Type_Descr  Fund_Code                 Fund_Code_Descr  \
16  Component Unit                       3047  SMITH-LEVER SPEC PRG GRNT FDF    
17  Component Unit                       3047  SMITH-LEVER SPEC PRG GRNT FDF    
18  Component Unit                       3047  SMITH-LEVER SPEC PRG GRNT FDF    
19  Component Unit                       3047  SMITH-LEVER SPEC PRG GRNT FDF    
20  Component Unit                       3047  SMITH-LEVER SPEC PRG GRNT FDF    

    Budget_Ref                                   Budget_Ref_Descr  ...  \
16        1330  SMITH-LEVER SPEC PRG GRNT FDF                 ...  ...   
17        1330  SMITH-LEVER SPEC PRG GRNT FDF                 ...  ...   
18        1330  SMITH-LEVER SPEC PRG GRNT FDF                 ...  ...   
19        1330  SMITH-LEVER SPEC PRG GRNT FDF                 ...  ...   
20        1330  SMITH-LEVER SPEC PRG GRNT FDF                 ...  ...   

    State_Function                               State_Function_Descr  \
16              26  Universities                                  ...   
17              26  Universities                                  ...   
18              26  Universities                                  ...   
19              26  Universities                                  ...   
20              26  Universities                                  ...   

    Primary_Account_Code Primary_Account_Code_Descr Intermediate_Account_Code  \
16    SALARIES AND WAGES         SALARIES AND WAGES                     51000   
17    SALARIES AND WAGES         SALARIES AND WAGES                     51000   
18    SALARIES AND WAGES         SALARIES AND WAGES                     51000   
19    SALARIES AND WAGES         SALARIES AND WAGES                     51000   
20    SALARIES AND WAGES         SALARIES AND WAGES                     51000   

   Intermediate_Account_Code_Descr           Detail_Account_Code  \
16  SALARIES AND WAGES                 UNEMPLOYMENT COMPENSATION   
17  SALARIES AND WAGES                      WORKERS COMPENSATION   
18  SALARIES AND WAGES              GROUP HEALTH HOSPITALIZATION   
19  SALARIES AND WAGES                                     OASDI   
20  SALARIES AND WAGES                                  MEDICARE   

         Detail_Account_Code_Descr     Total Unnamed: 21  
16  UNEMPLOYMENT COMPENSATION        108.820         nan  
17  WORKERS COMPENSATION             579.490         nan  
18  GROUP HEALTH HOSPITALIZATION   27218.570         nan  
19  OASDI                           7518.460         nan  
20  MEDICARE                        3139.450         nan  

[5 rows x 22 columns]

In [129]:
KUexpenditure["Primary_Account_Code"] = KUexpenditure["Primary_Account_Code_Descr"].str.strip()

In [125]:
KSUexpenditure.groupby(["Primary_Account_Code_Descr", "Detail_Account_Code_Descr"]).sum().sort_values("Total", ascending=False)

Fiscal_Year  \
Primary_Account_Code_Descr   Detail_Account_Code_Descr                     
SALARIES AND WAGES           UNCLASSIFIED REGULAR                1059450   
NONEXPENSE ITEMS             OTHER LOANS                           12108   
GRANTS CLAIMS SHARED REVENUE STUDENT SCHOLAR FELLOW GRANT         274448   
CAPITAL OUTLAY               BUILDINGS                            211890   
SALARIES AND WAGES           GROUP HEALTH HOSPITALIZATION        1019090   
                             EDUCATIONAL INST EE RETIRE SYS       956532   
GRANTS CLAIMS SHARED REVENUE FED AID TO QUALIFIED NON STATE        74666   
SALARIES AND WAGES           OASDI                               1077612   
CONTRACTUAL SERVICES         OTHER FEES                           700246   
SALARIES AND WAGES           UNCLASSIFIED TEMPORARY               554950   
                             STUDENT EMPLOYEES HOURLY             875812   
                             STUDENT RESEARCH ASSISTANTS          423780   
EXPENSE TRANSFERS            DEBT SVC TRANSFER PRINCIPAL           40360   
DEBT SERVICE                 BOND PRINCIPAL                        32288   
                             BOND INTEREST                         72648   
EXPENSE TRANSFERS            DEBT SVC TRANSFER INTEREST            76684   
CONTRACTUAL SERVICES         OTHER CONTRACTUAL SERVICES           508536   
                             ELECTRICITY                          195746   
COMMODITIES                  PRO SCIENTIFIC SUPPLY OTHER          738588   
SALARIES AND WAGES           UNCLASS HOLIDAY INCL IN BASE        1029180   
                             GRADUATE TEACHING ASSISTANTS         179602   
CONTRACTUAL SERVICES         REPAIR AND SVC BLDG GROUND           548896   
                             OTHER PROFESSIONAL FEES              603382   
NONEXPENSE ITEMS             OTHER NON EXPENSE ITEMS               74666   
CONTRACTUAL SERVICES         DATABASE ACCESS FEES NOT OITS        437906   
CAPITAL OUTLAY               PRO AND SCIENTIFIC EQUIP             221980   
EXPENSE TRANSFERS            INTERFUND TRANSFER OUT                84756   
SALARIES AND WAGES           PUBLIC EE RETIREMENT SYSTEM          774912   
GRANTS CLAIMS SHARED REVENUE ENROLL FEES SPONS INST GRANT         177584   
COMMODITIES                  FOOD FOR HUMAN CONSUMPTION           308754   
...                                                                  ...   
CAPITAL OUTLAY               DEPR LAND IMPROVEMENTS NON CAP         8072   
                             TELECOM TRANSMISSION NON CAP           6054   
CONTRACTUAL SERVICES         INTERGOV LD COMM SVC OP OITS           8072   
CAPITAL OUTLAY               PRINTING EQUIP NON CAPITAL            18162   
COMMODITIES                  DRUGS AND PHARMACEUTICALS             16144   
                             OTHER FUEL                            14126   
GRANTS CLAIMS SHARED REVENUE PROPERTY DAMAGE OR LOSS CLAIMS        10090   
COMMODITIES                  FUEL OIL DIESEL AND KEROSENE          24216   
GRANTS CLAIMS SHARED REVENUE ROYALTY PAYMENTS                      18162   
CONTRACTUAL SERVICES         COMPUTER EQUIPMENT RENTAL             22198   
COMMODITIES                  PROPANE BUTANE NATURAL GAS            32288   
CONTRACTUAL SERVICES         PRINT BIND SURCHARGE ST ONLY          22198   
CAPITAL OUTLAY               FCIP-PRINCIPAL-BI-C/I                  6054   
                             ASSESSMENTS FOR IMPROVEMENTS           4036   
CONTRACTUAL SERVICES         EMPLOYEE AWARDS NOT IRS REPORT        32288   
                             MOVING EE PERS EFFECTS IN ST           2018   
GRANTS CLAIMS SHARED REVENUE LOCAL FUNDS SHORTAGES                 22198   
CONTRACTUAL SERVICES         STATE CAR EXPENSE-INTERNATL           16144   
                             BANK FEES                             18162   
                             INTERGOV LD COMM SVC CAP OITS          8072   
                             COMP

In [126]:
KUexpen = KUexpenditure[KUexpenditure.Primary_Account_Code_Descr == 'SALARIES AND WAGES']

In [130]:
KUexpen

Empty DataFrame
Columns: [Fiscal_Year, Fiscal_Quarter, Business_Unit, Business_Unit_Descr, GAAP_Fund_Type, GAAP_Fund_Type_Descr, Fund_Code, Fund_Code_Descr, Budget_Ref, Budget_Ref_Descr, Program_Code, Program_Code_Descr, State_Function, State_Function_Descr, Primary_Account_Code, Primary_Account_Code_Descr, Intermediate_Account_Code, Intermediate_Account_Code_Descr, Detail_Account_Code, Detail_Account_Code_Descr, Total, Unnamed: 21]
Index: []

[0 rows x 22 columns]

<b>Answer:</b> KSU and KU's salaries and wages breakdown to unclassified regular, group health hospitalization, OASDI, Unclassified Temporary, unlcass holiday incl in base, public ee retirement system, calssified regular. 

### Q10: Compare the breakdown costs from Q9 side by side.
When you merge the two subsets, make sure to add the attribute <b>how ="outer"</b> so that you can keep every item from both sets.
*For example: merge = pd.merge(df1, df2, how="outer", on="columnname")

In [135]:
merged = pd.merge(KSUexpen, KUexpen, how="outer", on="Primary_Account_Code_Descr")

In [136]:
merged

Fiscal_Year_x  Fiscal_Quarter_x  Business_Unit_x  \
0              2018                 1              367   
1              2018                 1              367   
2              2018                 1              367   
3              2018                 1              367   
4              2018                 1              367   
5              2018                 1              367   
6              2018                 1              367   
7              2018                 1              367   
8              2018                 1              367   
9              2018                 1              367   
10             2018                 1              367   
11             2018                 1              367   
12             2018                 1              367   
13             2018                 1              367   
14             2018                 1              367   
15             2018                 1              367   
16             2018                 1              367   
17             2018                 1              367   
18             2018                 1              367   
19             2018                 1              367   
20             2018                 1              367   
21             2018                 1              367   
22             2018                 1              367   
23             2018                 1              367   
24             2018                 1              367   
25             2018                 1              367   
26             2018                 1              367   
27             2018                 1              367   
28             2018                 1              367   
29             2018                 1              367   
...             ...               ...              ...   
7750           2018                 4              367   
7751           2018                 4              367   
7752           2018                 4              367   
7753           2018                 4              367   
7754           2018                 4              367   
7755           2018                 4              367   
7756           2018                 4              367   
7757           2018                 4              367   
7758           2018                 4              367   
7759           2018                 4              367   
7760           2018                 4              367   
7761           2018                 4              367   
7762           2018                 4              367   
7763           2018                 4              367   
7764           2018                 4              367   
7765           2018                 4              367   
7766           2018                 4              367   
7767           2018                 4              367   
7768           2018                 4              367   
7769           2018                 4              367   
7770           2018                 4              367   
7771           2018                 4              367   
7772           2018                 4              367   
7773           2018                 4              367   
7774           2018                 4              367   
7775           2018                 4              367   
7776           2018                 4              367   
7777           2018                 4              367   
7778           2018                 4              367   
7779           2018                 4              367   

               Business_Unit_Descr_x  GAAP_Fund_Type_x  \
0     Kansas State University                       12   
1     Kansas State University                       12   
2     Kansas State University                       12   
3     Kansas State University                       12   
4     Kansas State University                       12   
5     Kansas State University                       12   
6  

<b>Answer:</b>

## III. BONUS QUESTIONS
* Each question's worth 15 points.


### B1: Which university made a profit in the fiscal year of 2018?

In [67]:
KSUrevenue.Total.sum()

854324643.84

In [69]:
KSUexpenditure.Total.sum()

1018584373.02

In [70]:
854324643.84-1018584373.02


-164259729.17999995

In [71]:
KUrevenue.Total.sum()

808767902.55

In [72]:
KUexpenditure.Total.sum()

990238875.6099999

In [73]:
808767902.55-990238875.6099999


-181470973.05999994

<b>Answer:</b> Neither of them made a profit

### B2: Which software program do you like the most in this course? Why?

<b>Answer:</b> I liked python the most because it is the one I understand the best and the online tutorial is helpful.

### END OF THE TEST